File: dt_brightlifecare_customization/public/js/material_request.js
frappe.ui.form.on('Material Request', {
    refresh: function (frm) {
        if (frm.doc.docstatus === 1) {
            if(frm.doc.material_request_type == "Manufacture"){
                frm.page.add_inner_button(__('Create MRP'), function () {
                    frappe.call({
                        method: 'dt_brightlifecare_customization.mrp.custom.material_request.material_request.create_mrp',
                        args: { material_request: frm.doc.name, use_defaults: 0 ,company: frm.doc.company},
                        callback: function (r) {
                            if (r.message) {
                                let mrp_doc = frappe.model.sync([r.message])[0];
                                frappe.set_route('Form', mrp_doc.doctype, mrp_doc.name);
                            }
                        },
                        error: function (err) {
                            frappe.throw({
                                title: __("Some fields are missing in items"),
                                indicator: "red",
                                message: err.message,
                            });
                        }
                    });
                });
            }
        }

        if (frm.doc.docstatus === 1 && frm.doc.custom_suppliers?.length && frm.doc.material_request_type == "Purchase") {
            frm.add_custom_button("Create RFQs", () => {
                frappe.call({
                    method: "dt_brightlifecare_customization.public.py.material_request.create_rfq_from_suppliers",
                    args: { docname: frm.doc.name },
                    callback(r) {
                        if (r.message) {
                            frappe.msgprint("RFQs Created: " + r.message.join(", "));
                        }
                    }
                });
            });
        }

        frm.remove_custom_button('Request for Quotation', 'Create');


        // setTimeout(() => {
        //     frm.page.actions.find('[data-label="Request for Quotation"]').parent().parent().remove();
        // }, 100);
    }
});






frappe.ui.form.on('Material Request Item', {
    item_code: function (frm, cdt, cdn) {
        let row = locals[cdt][cdn];

        if (!row.item_code) {
            frappe.model.set_value(cdt, cdn, 'custom_linked_supplier', '[]');
            return;
        }

        frappe.db.get_doc('Item', row.item_code)
            .then(doc => {
                if (doc && doc.supplier_items && doc.supplier_items.length > 0) {
                    let suppliers = doc.supplier_items
                        .filter(s => s.supplier)
                        .map(s => s.supplier);

                    let supplier_json = JSON.stringify(suppliers, null, 2);

                    frappe.model.set_value(cdt, cdn, 'custom_linked_supplier', supplier_json);
                } else {
                    frappe.model.set_value(cdt, cdn, 'custom_linked_supplier', '[]');
                }
            })
            .catch(() => {
                frappe.model.set_value(cdt, cdn, 'custom_linked_supplier', '[]');
            });
    }
});

File: dt_brightlifecare_customization/public/py/material_request.py
import frappe
from frappe.model.document import Document

def before_save(doc, method):
    supplier_set = set()
    doc.custom_suppliers = []

    for item in doc.items:
        if not item.item_code:
            continue

        item_doc = frappe.get_doc("Item", item.item_code)

        for supp in item_doc.supplier_items:
            if supp.supplier and supp.supplier not in supplier_set:
                supplier_set.add(supp.supplier)
                doc.append("custom_suppliers", {
                    "supplier": supp.supplier
                })







# Code to create RFQ for each supplier against their respective items
import frappe
from frappe.utils import nowdate


@frappe.whitelist()
def create_rfq_from_suppliers(docname):
    doc = frappe.get_doc("Material Request", docname)
    rfq_names = []

    # Find suppliers already used in RFQs for this Material Request
    existing_rfq_suppliers = frappe.get_all(
        "Request for Quotation Item",
        filters={"material_request": docname},
        fields=["parent"]
    )

    existing_suppliers = set()

    for rfq_item in existing_rfq_suppliers:
        suppliers = frappe.get_all(
            "Request for Quotation Supplier",
            filters={"parent": rfq_item.parent},
            fields=["supplier"]
        )
        for s in suppliers:
            existing_suppliers.add(s.supplier)

    supplier_items = {}

    for item in doc.items:
        item_doc = frappe.get_doc("Item", item.item_code)
        for supp in item_doc.supplier_items:
            if supp.supplier in [s.supplier for s in doc.custom_suppliers]:
                if supp.supplier in existing_suppliers:
                    continue  # Skip if already in RFQ for this material request
                supplier_items.setdefault(supp.supplier, []).append({
                    "item_code": item.item_code,
                    "qty": item.qty,
                    "schedule_date": item.schedule_date,
                    "warehouse": item.warehouse,
                    "conversion_factor": item.conversion_factor,
                    "uom": item.uom,
                    "material_request": doc.name,
                    "material_request_item": item.name
                })

    for supplier, items in supplier_items.items():
        rfq = frappe.new_doc("Request for Quotation")
        rfq.transaction_date = nowdate()  # Set today's date
        rfq.append("suppliers", {"supplier": supplier})
        for item in items:
            rfq.append("items", item)

        sup = frappe.get_doc("Supplier", supplier)
        if sup.custom_connected_users:
            for user_entry in sup.custom_connected_users:
                rfq.append("custom_connected_users", {
                    "user": user_entry.user
                    # add other fields if needed (e.g., role, email, etc.)
                })

                
        rfq.message_for_supplier = "Please provide your best quote."
        rfq.save()
        rfq_names.append(rfq.name)

    if not rfq_names:
        frappe.msgprint("RFQs already exist.")
    return rfq_names
